auth/src/main/java/io/apicurio/common/apps/auth/authn/AppAuthenticationMechanism.java

```diff
@@ -250,14 +261,39 @@ public String getAccessToken(Pair<String, String> clientCredentials, String cred
         OidcAuth oidcAuth = new OidcAuth(httpClient, clientCredentials.getLeft(), clientCredentials.getRight(), Duration.ofSeconds(1), scope.orElse(null));
         try {
             String jwtToken = oidcAuth.authenticate();//If we manage to get a token from basic credentials, try to authenticate it using the fetched token using the identity provider manager
-            cachedAccessTokens.put(credentialsHash, new WrappedValue<>(Duration.ofMinutes(accessTokenExpiration), Instant.now(), jwtToken));
+            cachedAccessTokens.put(credentialsHash, new WrappedValue<>(getAccessTokenExpiration(jwtToken), Instant.now(), jwtToken));
             return jwtToken;
         } catch (NotAuthorizedException | ForbiddenException ex) {
-            cachedAuthFailures.put(credentialsHash, new WrappedValue<>(Duration.ofMinutes(accessTokenExpiration), Instant.now(), ex));
+            cachedAuthFailures.put(credentialsHash, new WrappedValue<>(getAccessTokenExpiration(null), Instant.now(), ex));
             throw ex;
         }
     }
 
+    /**
+     * Figure out how long to cache a given JWT.  The token can be null (if authentication fails), in which case
+     * the configured default expiration time will be used.
+     */
+    protected Duration getAccessTokenExpiration(String jwtToken) {
+        if (jwtToken == null) {
+            return Duration.ofMinutes(accessTokenExpiration);
+        }
+        try {
+            JsonWebToken parsedToken = jwtParser.parseOnly(jwtToken);
+            
+            // Convert the expiration to an Instant, and subtract 10s (we want to stop using it 10s before it expires).
```

Comment:
The comment does not match the code, you're subtracting 30, not 10.

Reply:
Whoops! :)  And I suppose it should be configurable...